### DISCUSSION TITLE
Generate Multistage dockerFile and docker-compose

### DISCUSSION THREAD

**Discussion by bramlihamza** [eclipse/jkube#1708] [Q&A]
Hi First of all, thank you for this amazing project .
I want to generate a multi stage docker file with java and  using the DockerFileBuilder () but I can't, there is a solution or example  for that? 
It's possible to generate docker-compose yaml file? 
Thank you

**Comment by rohanKanojia** [comment]
Could you please elaborate your use case? Maybe share a reproducer project to provide us more insight on what you're trying?

Most of the Docker related code is inherited from Fabric8 Docker Maven Plugin. I do see Docker Compose support in DMP : https://dmp.fabric8.io/#docker-compose . But I haven't tried it myself with jkube.

**Reply by bramlihamza** [reply]
Hi, thank you for the fast response.

I want to reproduce some think like this .

```
# First stage:  **build environment**
FROM maven:3.5.0-jdk-8-alpine **AS builder**

# add pom.xml and source code
ADD ./pom.xml pom.xml
ADD ./src src/

# package jar
RUN mvn clean package

# Second **stage: minimal runtime environment**
From openjdk:8-jre-alpine

# copy jar from the first stage
COPY **--from=builder** target/my-app-1.0-SNAPSHOT.jar my-app-1.0-SNAPSHOT.jar

EXPOSE 8080

CMD ["java", "-jar", "my-app-1.0-SNAPSHOT.jar"]
```



So I want to create a **first stage to build** the project, then create a **second stage** to copy the artifact from the first stage and run it.

I tried to do this, but I was blocked in the second stage.


```
 new DockerFileBuilder()
        .basedir("/")
        .baseImage("**maven:3.5.0-jdk-8-alpine** ### **AS** **builder**")
        .expose(Collections.singletonList("8080"))
			/////
			/////
        .entryPoint(a)
        
        
        .write(javaProject.getWorkDir().toFile());
```

thank you .

**Reply by rohanKanojia** [reply]
Looks like our [DockerFileBuilder](https://github.com/eclipse/jkube/blob/master/jkube-kit/config/image/src/main/java/org/eclipse/jkube/kit/config/image/build/DockerFileBuilder.java) doesn't seem to support Multistage dockerFile. However it shouldn't be that hard to implement. 

Would it be possible for you to contribute this functionality? We are planning to release next version within a month.

**Reply by rohanKanojia** [reply]
Shall I convert this discussion into an issue?

**Comment by bramlihamza** [comment]
Hi Rohan, 

Sorry for the late response. 

I tried to add this functionality last week and it works fine for me.
I have created an issue and pushed my code. You can review it if you want https://github.com/eclipse/jkube/pull/1734. I'am open to your suggestions if you have an another implementation.


By the way I started to use this functionality for my graduation project right now and it works.
Thank you,
